File: PHP/One/Transforms/ResolveGenericTypeIdentifiers.php
<?php

namespace One\Transforms\ResolveGenericTypeIdentifiers;

use One\AstTransformer\AstTransformer;
use One\Ast\AstTypes\UnresolvedType;
use One\Ast\AstTypes\GenericsType;
use One\Ast\Types\Class_;
use One\Ast\Types\Method;
use One\Ast\Interfaces\IType;

class ResolveGenericTypeIdentifiers extends AstTransformer {
    function __construct() {
        parent::__construct("ResolveGenericTypeIdentifiers");
        
    }
    
    protected function visitType($type) {
        parent::visitType($type);
        
        //console.log(type && type.constructor.name, JSON.stringify(type));
        if ($type instanceof UnresolvedType && (($this->currentInterface instanceof Class_ && in_array($type->typeName, $this->currentInterface->typeArguments)) || ($this->currentMethod instanceof Method && in_array($type->typeName, $this->currentMethod->typeArguments))))
            return new GenericsType($type->typeName);
        
        return $type;
    }
}

File: CSharp/One/Transforms/ResolveGenericTypeIdentifiers.cs
using One.Ast;
using One;

namespace One.Transforms
{
    public class ResolveGenericTypeIdentifiers : AstTransformer
    {
        public ResolveGenericTypeIdentifiers(): base("ResolveGenericTypeIdentifiers")
        {
            
        }
        
        protected override IType visitType(IType type)
        {
            base.visitType(type);
            
            //console.log(type && type.constructor.name, JSON.stringify(type));
            if (type is UnresolvedType unrType && ((this.currentInterface is Class class_ && class_.typeArguments.includes(unrType.typeName)) || (this.currentMethod is Method meth && meth.typeArguments.includes(unrType.typeName))))
                return new GenericsType(unrType.typeName);
            
            return type;
        }
    }
}